src/cascadia/QueryExtension/OpenAILLMProvider.cpp

```diff
@@ -0,0 +1,125 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
+#include "pch.h"
+#include "OpenAILLMProvider.h"
+#include "../../types/inc/utils.hpp"
+#include "LibraryResources.h"
+
+#include "OpenAILLMProvider.g.cpp"
+
+using namespace winrt::Windows::Foundation;
+using namespace winrt::Windows::Foundation::Collections;
+using namespace winrt::Windows::UI::Core;
+using namespace winrt::Windows::UI::Xaml;
+using namespace winrt::Windows::UI::Xaml::Controls;
+using namespace winrt::Windows::System;
+namespace WWH = ::winrt::Windows::Web::Http;
+namespace WSS = ::winrt::Windows::Storage::Streams;
+namespace WDJ = ::winrt::Windows::Data::Json;
+
+static constexpr std::wstring_view applicationJson{ L"application/json" };
+static constexpr std::wstring_view acceptedModel{ L"gpt-3.5-turbo" };
+static constexpr std::wstring_view openAIEndpoint{ L"https://api.openai.com/v1/chat/completions" };
+
+namespace winrt::Microsoft::Terminal::Query::Extension::implementation
+{
+    void OpenAILLMProvider::SetAuthentication(const Windows::Foundation::Collections::ValueSet& authValues)
+    {
+        _AIKey = unbox_value_or<hstring>(authValues.TryLookup(L"key").try_as<IPropertyValue>(), L"");
+        _httpClient = winrt::Windows::Web::Http::HttpClient{};
+        _httpClient.DefaultRequestHeaders().Accept().TryParseAdd(applicationJson);
+        _httpClient.DefaultRequestHeaders().Authorization(WWH::Headers::HttpCredentialsHeaderValue{ L"Bearer", _AIKey });
+    }
+
+    void OpenAILLMProvider::ClearMessageHistory()
+    {
+        _jsonMessages.Clear();
+    }
+
+    void OpenAILLMProvider::SetSystemPrompt(const winrt::hstring& systemPrompt)
+    {
+        WDJ::JsonObject systemMessageObject;
+        winrt::hstring systemMessageContent{ systemPrompt };
+        systemMessageObject.Insert(L"role", WDJ::JsonValue::CreateStringValue(L"system"));
+        systemMessageObject.Insert(L"content", WDJ::JsonValue::CreateStringValue(systemMessageContent));
+        _jsonMessages.Append(systemMessageObject);
+    }
+
+    void OpenAILLMProvider::SetContext(const Extension::IContext context)
+    {
+        _context = context;
```

Comment:
This is where we'd ideally take the argument by-copy and then move it into place. This lets the caller move the instance into the argument which you then move into the member.
(A very minor nit.)

Reply:
Oh and: Moving only works if the argument is not const. If it's const it'll not fail to compile, because... C++ reasons. Instead, it'll silently just create a copy. Bu With high enough compiler warnings it'll warn you about it though, but idk if it's enabled for this project.